Web: Add plugin status types (okta) and add `error` kind to ToolTIp.tsx

kimlisa opened the pull request:
related: https://github.com/gravitational/teleport/pull/44431

🛑 do not merge: backend work is required that is TBD, it may change some fields in this PR

this PR contains just the necessary changes to implement the okta status page located in [enterprise](https://github.com/gravitational/teleport.e/pull/4784)

in the `integrations` list, only okta will have a `view status` link
<img width="871" alt="image" src="https://github.com/user-attachments/assets/4044e59d-e50c-4c37-9fd6-59fd6e5f213b">

added a `error` kind to tool tip
<img width="305" alt="image" src="https://github.com/user-attachments/assets/e4b62fe7-4c65-4767-94c3-bc37a6d4981b">

kimlisa committed:
Add error kind to tool tip

kimlisa committed:
Add status link to okta dropdown in the integraiton list

kimlisa committed:
Add a status route

kimlisa committed:
Add some types related plug status endpoint

ryanclark commented:
I feel like this option being hidden under "Options" would mean no one would know its there - can the row be clickable, or can we display "View Status" as a button and have a delete icon button too?

gzdunek commented:
Also a UX question: should we put error messages in tooltips?
Maybe we should just always show them on the page?

When there are sync errors [(screesnohst)](https://private-user-images.githubusercontent.com/43280172/353274709-a139fb33-7363-471d-9659-fda535fb986b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIzNTQ4OTYsIm5iZiI6MTcyMjM1NDU5NiwicGF0aCI6Ii80MzI4MDE3Mi8zNTMyNzQ3MDktYTEzOWZiMzMtNzM2My00NzFkLTk2NTktZmRhNTM1ZmI5ODZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzMwVDE1NDk1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhNDA3NTQzMDFjODZkMGQ0Nzc5NGU2MWYyMjVkMzRhMmNjNmI0NzNmNDZkM2ZlZmQxMzRjMWU0YzViMTRjZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.o7L7lf_Kgl8w8q1SkMfci0SiZCN-0PUFNUe3PxsyqY0), there are three warning icons. To see the actual problem, I have to click/hover them one by one (it feels inconvenient).

roraback commented:
> I feel like this option being hidden under "Options" would mean no one would know its there - can the row be clickable, or can we display "View Status" as a button and have a delete icon button too?

I agree. As we start to build out detail pages, I'd love for the entire row (or card) to become clickable so that users can navigate more easily.

roraback commented:
>Also a UX question: should we put error messages in tooltips?
>Maybe we should just always show them on the page?
>
>When there are sync errors, there are three warning icons. To see the actual problem, I have to click/hover them one by one (it feels inconvenient).

I hear you here, but trying to show errors inline can make the page look quite messy if we don't know what length we need to plan for. That said, I think it might be good to show _some_ error text along the lines of "Failed 3m ago" alongside the warning icon, and make the whole thing a tooltip target so you have a bigger area to hover than just a little square icon. 

What do you think about that @gzdunek @kimlisa ? It still requires hovering for the details, but it gives you a bit more insight on the recency and provides a bigger hover target.

gzdunek commented:
If we are concerned about the errors length, we can always try the good old text ellipsis:

https://github.com/user-attachments/assets/db17ab9f-3892-44bf-8a71-638bed1b87ea


The thing I don't like in tooltips is that often they have poor accessibility, especially on touch devices. For example, when it shows on hover, copying its content is impossible:

https://github.com/user-attachments/assets/35e2b20c-f582-4ab5-95b7-4dbe51c38c3d

But I'm not insisting on this change, I just wanted to add something to consider.

roraback commented:
> If we are concerned about the errors length, we can always try the good old text ellipsis:
I thought about this as well. The problem I see is that it still is problematic for the layout of a dashboard, particularly where you have two tiles side-by-side. If you expand the one tile, you're either also expanding the tile next to it with a lot of empty space or you're leaving a bunch of empty space _below_ that tile. If we did this kind of behavior, I'd want it to be like the Resources tiles when you overflow the labels—it expands but overlaps the next tile below it until you stop hovering, at which point it snaps back.

I hear you on the accessibility concern though... If copying the error message would be useful, we _could_ make it open on hover, but stay open on click so that if they click on it, it isn't dismissed until they click away.

gzdunek commented:
>I thought about this as well. The problem I see is that it still is problematic for the layout of a dashboard, particularly where you have two tiles side-by-side. If you expand the one tile, you're either also expanding the tile next to it with a lot of empty space or you're leaving a bunch of empty space below that tile. If we did this kind of behavior, I'd want it to be like the Resources tiles when you overflow the labels—it expands but overlaps the next tile below it until you stop hovering, at which point it snaps back.

I see this, thanks for explanation.

>I hear you on the accessibility concern though... If copying the error message would be useful, we could make it open on hover, but stay open on click so that if they click on it, it isn't dismissed until they click away.

Sounds good to me.

kimlisa committed:
Table add ons and remove ellipsis and re-order

- Allow on row click handler
- Allow custom row styler (func to conditionally apply styles)

kimlisa commented:
@ryanclark @gzdunek 

I made the following changes:
- allow rows that support status, to be clicked on and applies cursor style pointer https://github.com/gravitational/teleport/pull/44788/commits/ab982eb83066e922a4d1475704b1182a6256bcd6
- in enterprise, i made the tooltip sticky https://github.com/gravitational/teleport.e/pull/4784/commits/371cbc0216c1b8b6ed9f808f702ea9a9f4a73194

gzdunek commented on web/packages/design/src/DataTable/Table.tsx:
```diff
@@ -126,7 +127,15 @@ export function Table<T>({
+      rows.push(
+        <tr
+          key={rowIdx}
+          onClick={row?.onClick ? () => row.onClick(item) : undefined}
```
```suggestion
          onClick={() => row.onClick?.(item)}
```

kimlisa replied:
i did as you suggested, but it doesn't matter right that it's still technically calling a function (but nothings happening b/c it's undefined)

gzdunek commented on web/packages/design/src/DataTable/Table.tsx:
```diff
@@ -126,7 +127,15 @@ export function Table<T>({
+        <tr
+          key={rowIdx}
+          onClick={row?.onClick ? () => row.onClick(item) : undefined}
+          style={row?.onStyle ? row.onStyle(item) : undefined}
```
I think `getStyle` would make more sense here :)

gzdunek commented on web/packages/teleport/src/services/integrations/oktastatustypes.ts:
```diff
@@ -0,0 +1,93 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+// OktaPluginSyncStatusCode indicates the possible states of an Okta
+// synchronization service.
```
Can we convert these comments to JSDoc?

gzdunek commented on web/packages/teleport/src/services/integrations/oktastatustypes.ts:
```diff
@@ -0,0 +1,93 @@
+// OktaPluginSyncStatusCode indicates the possible states of an Okta
+// synchronization service.
+export enum PluginOktaSyncStatusCode {
+  // is the status code zero value,
```
I'd explicitly assign numbers to enum values.

gzdunek commented on web/packages/teleport/src/services/integrations/oktastatustypes.ts:
```diff
@@ -0,0 +1,93 @@
+  lastSuccess: Date;
+  lastFailed: Date;
+  numUsers: number;
+  /**
+   * Error contains a textual description of the reason the last synchronization
+   * failed. Only valid when StatusCode is OKTA_PLUGIN_SYNC_STATUS_CODE_ERROR.
+   */
```
We could enforce this via types:
```ts
export type OktaUserSyncDetails = {
  lastSuccess: Date;
  lastFailed: Date;
  enabled: boolean;
  numUsers: number;
} & (
  | {
      statusCode:
        | PluginOktaSyncStatusCode.Success
        | PluginOktaSyncStatusCode.Unspecified;
    }
  | {
      statusCode: PluginOktaSyncStatusCode.Error;
      /**
       * Contains a textual description of the reason the last synchronization
       * failed.
       */
      error: string;
    }
);
```
But I will leave it up to you, it may be a slight overkill for this case.

kimlisa replied:
left it as is

gzdunek commented on web/packages/teleport/src/services/integrations/oktastatustypes.ts:
```diff
@@ -0,0 +1,93 @@
+  appFilters: string[];
+  groupFilters: string[];
+  defaultOwners: string[];
+  error: string;
```
This error is also available only if status code is `OKTA_PLUGIN_SYNC_STATUS_CODE_ERROR`?

gzdunek commented on web/packages/teleport/src/services/integrations/oktastatustypes.ts:
> oktastatustypes.ts

I see the Go convention, but TBH I'm not sure if it should be applied to JS code. I'd stick with camelCase.

gzdunek replied:
Does the 👍 mean that you wanted to address it? :)

kimlisa replied:
renamed. that was so strange, i thought i renamed it

kimlisa committed:
Address CR

gzdunek commented on web/packages/teleport/src/Integrations/IntegrationList.tsx:
```diff
@@ -65,11 +66,27 @@ type Props<IntegrationLike> = {
+    history.push(cfg.getIntegrationStatusRoute(row.kind, row.name));
+  }
+
+  function handleOnRowStyle(row: IntegrationLike): React.CSSProperties {
```
```suggestion
  function getRowStyle(row: IntegrationLike): React.CSSProperties {
```

gzdunek commented on web/packages/teleport/src/Integrations/IntegrationList.tsx:
```diff
@@ -65,11 +66,27 @@ type Props<IntegrationLike> = {
 export function IntegrationList(props: Props<IntegrationLike>) {
+  const history = useHistory();
+
+  function handleOnRowClick(row: IntegrationLike) {
```
```suggestion
  function handleRowClick(row: IntegrationLike) {
```

kimlisa committed:
Rename file and func
